### PR TITLE
Refactor std::function usages from Brave Ads. 

### PR DESCRIPTION
**Pull request by aseren**
Refactor std::function usages from Brave Ads.

Next step is to consider changing Unretained(this) to weak ptr where needed: https://github.com/brave/brave-browser/issues/27973

Resolves https://github.com/brave/brave-browser/issues/26589

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- The changed code is covered by unit tests. No additional testing is required.

### CODE REVIEW

File: vendor/bat-native-ads/src/bat/ads/internal/account/account.cc [hunk 1]
```diff
@@ -126,15 +126,9 @@ void Account::Deposit(const std::string& creative_instance_id,
   }
 
   deposit->GetValue(
-      creative_instance_id, [=](const bool success, const double value) {
```
Comment (goodov) [inline vendor/bat-native-ads/src/bat/ads/internal/account/account.cc:129]: `[=]` is equal to capturing `this` by reference, so `base::Unretained(this)` is similar to the original implementation.

File: vendor/bat-native-ads/src/bat/ads/internal/account/user_data/conversion_user_data.cc [hunk 2]
```diff
@@ -5,27 +5,33 @@
...
 
-  builder::BuildConversion(creative_instance_id,
-                           [callback](const base::Value::Dict& user_data) {
-                             callback(user_data.Clone());
-                           });
+  builder::BuildConversion(
+      creative_instance_id,
+      base::BindOnce(
+          [](ConversionCallback callback, base::Value::Dict user_data) {
+            std::move(callback).Run(user_data.Clone());
```
Comment (goodov) [inline vendor/bat-native-ads/src/bat/ads/internal/account/user_data/conversion_user_data.cc:32]: you now have `base::Value::Dict user_data` here, so you don't need to `Clone()`.
`Run(std::move(user_data))`

File: vendor/bat-native-ads/src/bat/ads/internal/account/account_util_unittest.cc [hunk 3]
```diff
@@ -64,7 +64,7 @@ TEST_F(BatAdsAccountUtilTest, ResetRewards) {
   privacy::GetUnblindedPaymentTokens()->AddTokens(unblinded_payment_tokens);
 
   // Act
-  ResetRewards([](const bool success) {
+  ResetRewards(base::BindOnce([](const bool success) {
```
Comment (goodov) [inline vendor/bat-native-ads/src/bat/ads/internal/account/account_util_unittest.cc:67]: maybe it's a task for another PR, but we should replace such callbacks with `MockCallback`. Right now it's impossible to detect if the callback was actually called or not. IMO, this is critical, especially in such PRs when we refactor the callback foundation.

it should become something like this:
```
base::MockCallback<ResetRewardsCallback> callback;
EXPECT_CALL(callback, Run(true)).WillOnce([](const bool success) {
   ...
});
ResetRewards(callback.Get());
```

File: vendor/bat-native-ads/src/bat/ads/internal/account/user_data/user_data_builder_interface.h [hunk 4]
```diff
@@ -6,13 +6,13 @@
...
-#include <functional>
-
+#include "base/functional/callback_forward.h"
 #include "base/values.h"
 
 namespace ads {
 
-using UserDataBuilderCallback = std::function<void(const base::Value::Dict&)>;
+using UserDataBuilderCallback =
+    base::OnceCallback<void(const base::Value::Dict&)>;
```
Comment (goodov) [inline vendor/bat-native-ads/src/bat/ads/internal/account/user_data/user_data_builder_interface.h:15]: can this be by value?

File: vendor/bat-native-ads/src/bat/ads/internal/account/account.cc [hunk 5]
```diff
@@ -157,31 +151,50 @@ void Account::MaybeGetIssuers() const {
...
+  }
+
+  ProcessDeposit(creative_instance_id, ad_type, confirmation_type, value);
+}
+
 void Account::ProcessDeposit(const std::string& creative_instance_id,
                              const AdType& ad_type,
                              const ConfirmationType& confirmation_type,
                              const double value) const {
   transactions::Add(
```
[the remark this answers is not in the document]
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/account/account.cc:171]: Discussed in DM, thanks

File: vendor/bat-native-ads/src/bat/ads/internal/account/user_data/conversion_user_data.cc (hunk 2 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/account/user_data/conversion_user_data.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/account/user_data/user_data_builder_interface.h (hunk 4 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/account/user_data/user_data_builder_interface.h]: changed to passing by value

File: vendor/bat-native-ads/src/bat/ads/internal/account/account_util_unittest.cc (hunk 3 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/account/account_util_unittest.cc:67]: filed the issue: https://github.com/brave/brave-browser/issues/28056

**Review by goodov — APPROVED** [review]

**Review by tmancey — APPROVED** [review]
LGTM++

**Pull request merged**